### PR TITLE
Add return type `isTraitable`

### PR DESCRIPTION
**Pull request by ruudk**
I'd like to do this for more methods that are exposed in the compiled code. This will ease static analysis as there are less errors that are coming from Twig. 

This is just a first small PR to see if this is acceptable 😊

### CODE REVIEW

**Comment by stof** [comment]
I suggest also adding the `@return` phpdoc in 3.x

**Comment by ruudk** [comment]
I can change the PR, target v3, and then start applying `@return` types. Then in v4 we can (or not) change them to native types.

**Comment by stof** [comment]
@ruudk I would change them to native type in v4 (as done for the Twig public API)

**Comment by ruudk** [comment]
Alright, then we start with this PR for v4. When it's merged, I'll continue for v3 with `@return` tags, and then later the native types in v4 again.

**Comment by fabpot** [comment]
> Alright, then we start with this PR for v4. When it's merged, I'll continue for v3 with `@return` tags, and then later the native types in v4 again.

Let's start with v3, it makes things easier for me and more logical.

**Comment by ruudk** [comment]
Alright done. Targeting v3 and rebased.

~After this is merged, I will do a slightly bigger PR that changes a few more places in one go.~

**Comment by ruudk** [comment]
Made some changes. Turns out that we need to add the missing type hints to the `abstract Template`. 

Next to that, we now also type the 2 properties in the compiled templates.

File: src/Node/ModuleNode.php [hunk 1]
```diff
@@ -170,7 +171,13 @@ protected function compileClassHeader(Compiler $compiler)
             ->raw(" extends Template\n")
             ->write("{\n")
             ->indent()
+            ->write("/**\n")
+            ->write(" * @var Source\n")
+            ->write(" */\n")
             ->write("private \$source;\n")
```
Comment (fabpot) [inline src/Node/ModuleNode.php]: What about using a proper type hint (this class is final anyway and the variable private)?

```suggestion
            ->write("private Source \$source;\n")
```

File: src/Node/ModuleNode.php [hunk 2]
```diff
@@ -170,7 +171,13 @@ protected function compileClassHeader(Compiler $compiler)
...
+            ->write("/**\n")
+            ->write(" * @var array<string, Template>\n")
+            ->write(" */\n")
             ->write("private \$macros = [];\n\n")
```
Comment (fabpot) [inline src/Node/ModuleNode.php]:
```suggestion
            ->write("private array \$macros = [];\n\n")
```

File: src/Template.php [hunk 3]
```diff
@@ -107,11 +107,17 @@ public function getParent(array $context)
...
+     */
     protected function doGetParent(array $context)
     {
         return false;
     }
 
+    /**
+     * @return bool
+     */
     public function isTraitable()
```
Comment (fabpot) [inline src/Template.php]: As this class is internal, I think we can use a proper tye hint:

```suggestion
    public function isTraitable(): bool
```

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by ruudk** [comment]
I applied the feedback. I added more return types, since the `Template` class is `@internal` anyway. 

What do you think?

File: src/Node/ModuleNode.php [hunk 4]
```diff
@@ -115,7 +115,7 @@ protected function compileGetParent(Compiler $compiler)
         $parent = $this->getNode('parent');
 
         $compiler
-            ->write("protected function doGetParent(array \$context)\n", "{\n")
+            ->write("protected function doGetParent(array \$context): bool | string | Template | TemplateWrapper\n", "{\n")
```
Comment (fabpot) [inline src/Node/ModuleNode.php]: I think we don't add whitespace between the pipes (in Symfony I mean)

```suggestion
            ->write("protected function doGetParent(array \$context): bool|string|Template|TemplateWrapper\n", "{\n")
```

File: src/Template.php [hunk 5]
```diff
@@ -107,12 +103,12 @@ public function getParent(array $context)
         return $this->parents[$parent];
     }
 
-    protected function doGetParent(array $context)
+    protected function doGetParent(array $context) : bool | string | self | TemplateWrapper
```
Comment (fabpot) [inline src/Template.php]:
```suggestion
    protected function doGetParent(array $context): bool|string|self|TemplateWrapper
```

File: src/Template.php [hunk 6]
```diff
@@ -107,12 +103,12 @@ public function getParent(array $context)
...
     {
         return false;
     }
 
-    public function isTraitable()
+    public function isTraitable() : bool
```
Comment (fabpot) [inline src/Template.php]:
```suggestion
    public function isTraitable(): bool
```

File: src/Template.php [hunk 7]
```diff
@@ -52,24 +52,20 @@ public function __construct(Environment $env)
...
+    abstract public function getTemplateName() : string;
 
     /**
      * Returns debug information about the template.
      *
-     * @return array Debug information
+     * @return array<int, int> Debug information
      */
-    abstract public function getDebugInfo();
+    abstract public function getDebugInfo() : array;
```
Comment (fabpot) [inline src/Template.php]:
```suggestion
    abstract public function getDebugInfo(): array;
```

File: src/Template.php [hunk 8]
```diff
@@ -52,24 +52,20 @@ public function __construct(Environment $env)
 
     /**
      * Returns the template name.
-     *
-     * @return string The template name
      */
-    abstract public function getTemplateName();
+    abstract public function getTemplateName() : string;
```
Comment (fabpot) [inline src/Template.php]:
```suggestion
    abstract public function getTemplateName(): string;
```

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Template.php (hunk 8 above)
Comment (ruudk) [reply src/Template.php]: How come this is not automatically enforced by PHP CS Fixer or fabbot? I feel your valuable time can be used for other things 😁

**Comment by ruudk** [comment]
Applied the changes.

**Comment by fabpot** [comment]
Thank you @ruudk.

**Pull request merged**